trame_server/utils/typed_state.py

```diff
@@ -0,0 +1,501 @@
+    encoded state value.
+
+    :param state: Trame State which will be mutated / read from.
+    :param state_id: Associated trame string id where the data will be pushed / read from.
+    :param name: Name of the source field.
+    :param field_type: Type of the source field.
+    :param default: Default value of the source field.
+    :param default_factory: Default factory of the source field.
+    :param state_encoder: Encoder/decoder class for the proxy.
+    """
+
+    def __init__(
+        self,
+        *,
+        state: State,
+        state_id: str,
+        name: str,
+        field_type: type,
+        default,
+        default_factory,
+        state_encoder: IStateEncoderDecoder,
+    ):
+        self._state = state
+        self._state_id = state_id
+        self._name = name
+        self._default = default
+        self._encoder = state_encoder
+        self._type = field_type
+
+        # Set the default value to trame state if needed
+        default_value = default
+        if default_value == MISSING and default_factory != MISSING:
+            default_value = default_factory()
+        if default_value != MISSING:
+            self._state.setdefault(self._state_id, self._encoder.encode(default_value))
+
+    def __get__(self, instance, owner):
+        return self.get_value()
+
+    def __set__(self, instance, value):
+        self.set_value(value)
+
+    def get_value(self):
+        value = self._state[self._state_id]
+        return self._encoder.decode(value, self._type)
+
+    def set_value(self, value):
+        self._state[self._state_id] = self._encoder.encode(value)
+
+
+class _NameField:
+    """
+    Descriptor for fields to state id string equivalent.
+
+    :param state_id: Associated trame string id where the data will be pushed / read from.
+    """
+
+    def __init__(self, state_id: str):
+        self._state_id = state_id
+
+    def __get__(self, instance, owner):
+        return self._state_id
+
+
+class TypedState(Generic[T]):
+    """
+    Helper to have access to, mutate, and be notified of state changes using a strongly typed dataclass interface.
+
+    TypedState provides a type-safe wrapper around the trame State object, allowing to:
+    - Access and modify state using dataclass field names with full type hints
+    - Bind change callbacks to specific fields or combinations of fields
+    - Automatically handle encoding/decoding of complex types (enums, UUIDs, dates, etc.)
+    - Use namespaces to avoid conflicts between different state objects
+    """
+
+    _STATE_PROXY_DATACLASS_TYPE = "__state_proxy_dataclass_type"
+    _STATE_PROXY_FIELD_DICT = "__state_proxy_field_dict"
+    _STATE_PROXY_STATE_ID = "__state_proxy_state_id"
+
+    def __init__(
+        self,
+        state: State,
+        dataclass_type: Type[T],
+        *,
+        state_namespace="",
+        state_encoders: list[IStateEncoderDecoder] | None = None,
+    ):
```

Comment:
Small detail, but since you are defining the TypedState class, I would skip the `state_` prefix for __namespace__ and __encoders__.

Reply:
Sounds good, I removed the prefix!